fix: Addon collection and Library fetch on user login raise flags

elpiel opened the pull request:
## API call to retrieve Addons Collection failed on login:

![image](https://github.com/Stremio/stremio-core/assets/8925621/0010f775-98d4-4161-abf0-4b71af3ced24)

## API call to retrieve User Library failed on login:

![image](https://github.com/Stremio/stremio-core/assets/8925621/37514913-5083-4890-a235-8f46724a36f4)


## Both requests failing on login:

![image](https://github.com/Stremio/stremio-core/assets/8925621/ac36105a-db91-46d3-ab84-de30b231a47e)

elpiel committed:
fix: Addon collection and Library fetch on user login raise flags

Signed-off-by: Lachezar Lechev <lachezar@ambire.com>

elpiel committed:
fix: profile (de)serialize tests

Signed-off-by: Lachezar Lechev <lachezar@ambire.com>

elpiel committed:
chore: add migration for Profile

Signed-off-by: Lachezar Lechev <lachezar@ambire.com>

elpiel committed:
chore: add default for addons_locked and library_missing

Signed-off-by: Lachezar Lechev <lachezar@ambire.com>

elpiel committed:
feat: emit Events for addons locked and missing library updates

Signed-off-by: Lachezar Lechev <lachezar@ambire.com>

elpiel committed:
fix: migration test

Signed-off-by: Lachezar Lechev <lachezar@ambire.com>

elpiel committed:
chore: remove tracing log for CtxAuthResult

Signed-off-by: Lachezar Lechev <lachezar@ambire.com>

unclekingpin commented on src/runtime/msg/internal.rs:
```diff
@@ -23,7 +23,19 @@ pub type CtxStorageResponse = (
+#[derive(Debug)]
+pub struct CtxAuthResponse {
+    pub auth: Auth,
+    pub addons: Vec<Descriptor>,
```
would be better to replace the flags with 
```pub addons: Result<Vec<>, Error>```

the same is valid for library_items too

~Enum should be called CtxAuthResult. It is used in action called Internal::CtxAuthResult~

unclekingpin commented on src/models/ctx/ctx.rs:
```diff
@@ -229,66 +264,79 @@ fn authenticate<E: Env + 'static>(auth_request: &AuthRequest) -> Effect {
+            // lock if the result from fetching the addons has failed
+            let addons_locked = addon_collection_result.is_err();
+            // set the flag to true if fetching of the library has failed
+            let library_missing = datastore_library_result.is_err();
```
i dont like it that errors are swallowed here
if you change the CtxAuthResponse struct as i suggested this could be handled like

  future::try_join(addon_collection_fut, datastore_library_fut)
                    .then(move |(addons_result, library_items_result)| (auth, addons_result, library_items_result))
                    
then fallback to Official_addons somwhere else, not inside the authenticate function

elpiel replied:
Resolved

unclekingpin commented on src/types/profile/profile.rs:
```diff
@@ -15,10 +15,20 @@ pub type UID = Option<String>;
+    pub addons_locked: bool,
+    /// This missing library flag is raised when the API Library Collection fetch request on login
+    /// has failed to indicate why the user has an empty Library
+    #[serde(default)]
```
Do we need migration if we have serde default?

elpiel replied:
I preferred it for consistency.

elpiel committed:
fix: address PR review comments:
- Keep the results instead of discarding errors on authentication
- add error logs
- Profile - remove library_missing and update migration

Signed-off-by: Lachezar Lechev <lachezar@ambire.com>

unclekingpin commented on src/runtime/msg/internal.rs:
```diff
@@ -23,7 +23,12 @@ pub type CtxStorageResponse = (
+#[derive(Debug)]
+pub struct CtxAuthResponse {
+    pub auth: Auth,
+    pub addons_result: Result<Vec<Descriptor>, CtxError>,
```
those `_result` suffixes are fine but only if we use them everywhere, on every Result and on every Option, for consistency.
might be better to get rid of em

unclekingpin commented on src/models/ctx/ctx.rs:
```diff
@@ -229,66 +271,80 @@ fn authenticate<E: Env + 'static>(auth_request: &AuthRequest) -> Effect {
     let auth_api = APIRequest::Auth(auth_request.clone());
 
     EffectFuture::Concurrent(
-        E::flush_analytics()
```
I cannot check whats actually changed in the authenticate function. It seems ok, but ideally it shouldve been only the future join at the bottom.
```
    future::join(addon_collection_fut, datastore_library_fut).then(
        move |(addons, library_items)| {
            future::ok(CtxAuthResponse {
                auth,
                addons,
                library_items,
            })
        },
    )
```

unclekingpin reviewed (commented):
Migration is unnessesary but it wont cause any issues.
LGTM

elpiel committed:
chore: remove  mutable Profile from update_library

Signed-off-by: Lachezar Lechev <lachezar@ambire.com>

elpiel committed:
Merge branch 'development' into fix/user-unable-to-login-if-addon-collection-fetch-fails

elpiel committed:
fix: profile locking of addons
- lock/unlock addons on failed request after login
- check addon lock status on UpgradeAddon

Signed-off-by: Lachezar Lechev <lachezar@ambire.com>

tymmesyde commented on src/models/ctx/update_profile.rs:
```diff
@@ -119,6 +121,10 @@ pub fn update_profile<E: Env + 'static>(
         }
         Msg::Action(Action::Ctx(ActionCtx::UpgradeAddon(addon))) => {
+            if profile.addons_locked {
+                return addon_install_error_effects(addon, OtherError::UserAddonsAreLocked);
```
```suggestion
                return addon_upgrade_error_effects(addon, OtherError::UserAddonsAreLocked);
```

tymmesyde commented on src/models/ctx/update_profile.rs:
```diff
@@ -153,6 +159,10 @@ pub fn update_profile<E: Env + 'static>(
         }
         Msg::Internal(Internal::UninstallAddon(addon)) => {
+            if profile.addons_locked {
+                return addon_install_error_effects(addon, OtherError::UserAddonsAreLocked);
```
```suggestion
                return addon_uninstall_error_effects(addon, OtherError::UserAddonsAreLocked);
```

elpiel committed:
chore: update_profile applied PR review comment

Co-authored-by: Tim <tymmesyde@gmail.com>

elpiel committed:
chore: update_profile applied PR review comment

Co-authored-by: Tim <tymmesyde@gmail.com>
